# CMSE 202 Midterm (Section 003 - Fall 2025)

The goal of this midterm is to give you the opportunity to test out some of the skills that you've developed thus far this semester. In particular, you'll practice setting up a GitHub repository, committing and pushing repository changes, downloading data with command line tools, performing some data analysis, possibly using a new Python package, and writing a python class. You should find that you have all of the skills necessary to complete this exam with even just eight weeks of CMSE 202 under your belt!

You are encouraged to look through the entire exam before you get started so that you can appropriately budget your time and understand the broad goals of the exam. Once you've read through it, try doing Parts 0 and 1 first so that you have your repository set up and you download all necessary data files as they will be necessary to complete the assigned tasks. Let your instructor know right away if you have problems downloading the data!

The exam is set up so that even if you get stuck on one part there are opportunities to get points on the other parts, so consider jumping ahead if you feel like you aren't making progress and then come back later if you have time. If you have any questions during the exam, you may ask the instructor, the TA, or the LA privately: simply raise your hand and one of us will come over to you. 


**Important note about using online resources**: This exam is "open internet". That means that you can look up documentation, google how to accomplish certain Python tasks, etc. Being able to effectively use the internet for computational modeling and data science is a very important skill, so we want to make sure you have the opportunity to exercise that skill. **However: The use of any person-to-person communication software or generative AI tools is absolutely not acceptable.** If you are seen accessing your email, using a collaborative cloud storage or document software (e.g. Slack, Google Documents), or generative AIs (e.g. ChatGPT), you will be at risk for receiving a zero on the exam. Google often does give an AI summary regardless of search, so to avoid misusing AI on the exam you should not copy-paste any of the questions from the exam directly into Google. 

**Keep your eyes on your screen!** Unfortunately, there isn't enough space in the room for everyone to sit at their own table so please keep your eyes on your own screen. This exam is designed to give *you* the opportunity to show the instructor what you can do and you should hold yourself accountable for maintaining a high level of academic integrity. If any of the instructors observe suspicious behavior, you will, again, risk receiving a zero.

**Github Issues:** If at any point you encounter an issue with Github, such as a merge conflict, rather than spending your limited exam time trying to fix it, just write the code you would use to achieve the task in the notebook and move on. If you note the issue you encountered and have correct code in your notebook, you can still receive full credit. 

---
## Part 0: Academic integrity statement

Read the following statement and edit the markdown text to put your name in the statement. This is your commitment to doing your own authentic work on this exam.

> I, **Yan Pei**, affirm that this exam represents my own authetic work, without the use of any unpermitted aids or generative AIs or person-to-person communication. I understand that this exam an an opportunity to showcase my own progress in developing and improving my computational skills and have done my best to demonstrate those skills.

--
## Part 1: Add to your Git repository to track your progress on your exam (10 points total)

Before you get too far along in the exam, you're going to add it to the `cmse202-f25-turnin` repository you created in class (or in Homework 1) so that you can track your progress on the exam and preserve the final version that you turn in. In order to do this you need to

1. Navigate to your `cmse202-f25-turnin` repository and create a new directory called `midterm`.
2. Move this notebook into that **new directory** in your repository, then **add it and commit it to your repository**.
3. Finally, to test that everything is working, "git push" the file so that it ends up in your GitHub repository.

**Important**: Double check you've added your Professor and your TA as collaborators to your "turnin" respository (you should have done this in the previous homework assignment).

**Important**: Make sure that the version of this notebook that you are working on is the same one that you just added to your repository! If you are working on a different copy of the noteobok, **none of your changes will be tracked**!

The file should now show up on your GitHub account in the `cmse202-f25-turnin` repository inside the `midterm` directory you just created.  Periodically, **you'll be asked to commit your changes to the repository and push them to the remote GitHub location**. It can be good to get into a habit of committing your changes any time you make a significant modification.

&#9989; **Question 1.1 Put all the Git commands you used to add your midterm folder and file to your GitHub repository (4 points)**:

In [2]:
### Put your code here:
cd cmse-202-turnin
mkdir midterm
mv ~/Downloads/CMSE202-Midterm-Section003-STUDENT-F25.ipynb midterm/

SyntaxError: invalid syntax (170509374.py, line 2)

&#9989; **Question 1.2 Put the code the instructor will need to clone your repository here (1 points)**:


In [1]:
### Put your code here:
!git clone https://github.com/Rockypei/cmse202-f25-turnin.git

SyntaxError: invalid syntax (2186518450.py, line 2)

&#9989; **Question 1.3 Create a local branch called `exam_1_update`. What sequence of Git commands would you use to merge `exam_1_update` into `main`, resolve any conflicts if needed, and push the final result to GitHub?
 (3 points)**:


In [4]:
git checkout main
git pull origin main
git checkout -b exam_1_update
git add -A
git commit -m "Exam: updates on Part X (QY) in midterm notebook"
git push -u origin exam_1_update
git checkout main
git pull origin main
git merge --no-ff exam_1_update

SyntaxError: invalid syntax (3911078954.py, line 1)

&#9989; **Question 1.4 (2 points)**: After completing each of Part 2, Part 3, Part 4, and Part 5 below, save your notebook, commit the changes to your Git repository using a descriptive commit message, and push the changes to GitHub. Note that there are large &#128721; **STOP** signs reminding you to do this after each part. **Note:** If Github issues are encountered, just write the code you would use to commit and push changes under each of the stop signs. 

---
## Part 2: Downloading and analyzing unfamiliar data (12 points)



The data set at this web link `https://raw.githubusercontent.com/NkechiC/Data-files-for-teaching/refs/heads/main/202/heart_data.csv` contains clinical test results of 303 patients at the Cleveland Clinic in Cleveland, Ohio. This data is used to estimate the incidence of vascular conditions among the patients.​

&#9989; **Question 2.1  Use the `curl` function we learned in class to download the `heart_data.csv` dataset to your working directory. Include the code you used to download the dataset below.  (1 point)**:


In [3]:
### Put your code here:

age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 11325  100 11325    0     0  28162      0 --:--:-- --:--:-- --:--:-- 28383



63,1,3,145,233,1,0,150,0,2.3,0,0,1,1
37,1,2,130,250,0,1,187,0,3.5,0,0,2,1
41,0,1,130,204,0,0,172,0,1.4,2,0,2,1
56,1,1,120,236,0,1,178,0,0.8,2,0,2,1
57,0,0,120,354,0,1,163,1,0.6,2,0,2,1
57,1,0,140,192,0,1,148,0,0.4,1,0,1,1
56,0,1,140,294,0,0,153,0,1.3,1,0,2,1
44,1,1,120,263,0,1,173,0,0,2,0,3,1
52,1,2,172,199,1,1,162,0,0.5,2,0,3,1
57,1,2,150,168,0,1,174,0,1.6,2,0,2,1
54,1,0,140,239,0,1,160,0,1.2,2,0,2,1
48,0,2,130,275,0,1,139,0,0.2,2,0,2,1
49,1,1,130,266,0,1,171,0,0.6,2,0,2,1
64,1,3,110,211,0,0,144,1,1.8,1,0,2,1
58,0,3,150,283,1,0,162,0,1,2,0,2,1
50,0,2,120,219,0,1,158,0,1.6,1,0,2,1
58,0,2,120,340,0,1,172,0,0,2,0,2,1
66,0,3,150,226,0,1,114,0,2.6,0,0,2,1
43,1,0,150,247,0,1,171,0,1.5,2,0,2,1
69,0,3,140,239,0,1,151,0,1.8,2,2,2,1
59,1,0,135,234,0,1,161,0,0.5,1,0,3,1
44,1,2,130,233,0,1,179,1,0.4,2,0,2,1
42,1,0,140,226,0,1,178,0,0,2,0,2,1
61,1,2,150,243,1,1,137,1,1,1,0,2,1
40,1,3,140,199,0,1,178,1,1.4,2,0,3,1
71,0,1,160,302,0,1,162,0,0.4,2,2,2,1
59,1,2,150,212,1,1,157,0,1.6,2,0,2,1
51,1,2,110

&#9989; **Question 2.2 Read in the dataset as a Pandas data frame and then display the first 5 rows. (1 point)**: 



In [24]:
### Put your code here:



&#9989; **Question 2.3 Describe the dataset by finding all the summary statistics (1 point)**: 



In [26]:
### Put your code here:



&#9989; **Question 2.4 Now create a slice of the dataset that includes only the numerical variables `age`, `trestbps`, `chol`, `oldpeak`, and `thalach` (2 points)**: 



In [17]:
### Put your code here

&#9989; **Question 2.5 Make a grid of four (4) scatterplots to analyze the respective correlations between age, `age`,  and the rest of the variables: resting blood pressure, cholesterol level, stress test for depression, and maximum heart rate, as measured by the variables `trestbps`, `chol`, `oldpeak`, and `thalach`, respectively. Make sure to include labels and a title for your plots. (4 points)**: 


In [30]:
### Put your code here



&#9989; **Question 2.6 Describe the correlation relationships you observe between each pair of variables from your plots above (2 points)**: 


In [32]:
### Put your code here



---
### &#128721; STOP (1 point)
**Pause to commit your changes to your Git repository!**

Take a moment to save your notebook, commit  and push the changes to your Git repository.

---

## Part 3: Working with a new Python package (7 points)

[`Polars`](https://pola.rs/) is an open-source Python library for data manipulation. In this section, you will be exploring the `Polars` Python package along with its documentation [Docs](https://docs.pola.rs/). 


&#9989; **Question 3.1 Use pip install to install the Polars package. Include the code you used to install it here. Remember to include the --user flag if using JupyterHub. (1 point):** 

In [34]:
# Put your code here



&#9989; **Question 3.2 Now import the same `heart_data.csv` as a dataframe using Polars. (1 point):** 


In [2]:
# Put your code here


&#9989; **Question 3.3 Now use Polar to filter the dataframe and create a subset containing only entries with `age` higher than 55 years old. Display the first 5 rows of the new dataframe. (3 points):** 


In [38]:
# Put your code here



&#9989; **Question 3.4 Now write the subset dataframe above to a csv file called `heart_data_sub.csv`. (1 point):** 


In [40]:
# Put your code here



---
### &#128721; STOP (1 point)
**Pause to commit your changes to your Git repository!**

Take a moment to save your notebook, commit  and push the changes to your Git repository.

---

## Part 4: Writing a simple Python class (19 points)

In this section, we will be developing a Python class called `GradebookManager` for managing and manipulating student grade data from multiple CSV files, using the Pandas library.

&#9989; **Do this:** Begin by downloading the files from the following three weblinks using the `curl` command:
1. `https://raw.githubusercontent.com/NkechiC/Data-files-for-teaching/refs/heads/main/202/final_grades.csv`
2. `https://raw.githubusercontent.com/NkechiC/Data-files-for-teaching/refs/heads/main/202/midterm.csv`
3. `https://raw.githubusercontent.com/NkechiC/Data-files-for-teaching/refs/heads/main/202/quiz_grades.csv`

&#9989; **Question 4.1 Below is a half-completed GradebookManager class. Complete the code for the `init`, `load_data`, `merge_datasets`, `calculate_final_grades`, and `get_student_grades` methods. Follow the descriptions in the relevant DocString and the "COMPLETE THIS" comments. (12 points):** 


In [42]:
### Put your code here

import pandas as pd

class GradebookManager:
    """
    A class to manage, merge, and process student grade data from multiple CSV files.

    Attributes:
        dataframes (dict): A dictionary of loaded DataFrames, keyed by dataset name.
        merged_data (pd.DataFrame or None): The combined DataFrame after merging datasets.
    """

    def __init__(self, file_paths):
        """
        Initializes the GradebookManager with a list of file paths.

        Args:
            file_paths (list of str): Paths to CSV files containing student grade data.
        """
        ... ##COMPLETE THIS PART!

        
    def load_data(self, file_paths):
        """
        Loads CSV files into pandas DataFrames and stores them in a dictionary.

        Args:
            file_paths (list of str): Paths to CSV files.

        Raises:
            FileNotFoundError: If a file path is invalid.
            pd.errors.EmptyDataError: If a CSV file is empty.
        """
        for i, path in enumerate(file_paths):
            try:
                df = ... #COMPLETE THIS: load csv with Pandas
                self.dataframes[f'dataset_{i+1}'] = df
            
            except ...: #COMPLETE THIS LINE!
                print(f"Error: File not found at {path}")
            
            except ...: #COMPLETE THIS LINE!
                print(f"Error: Empty CSV file at {path}")

   
    def merge_datasets(self, on_column='student_id', how='outer'):
        """
        Merges all loaded datasets into a single DataFrame on a common column.

        Args:
            on_column (str): The column name to merge on (default is 'student_id').
            how (str): Type of merge to perform (default is 'outer'). 
                       Options: 'inner', 'outer', 'left', 'right'.

        Returns:
            None
        """
        if not self.dataframes:
            ... #COMPLETE THIS LINE!
            return

        first_df_key = list(self.dataframes.keys())[0]
        self.merged_data = self.dataframes[first_df_key]

        for key, df in list(self.dataframes.items())[1:]:
            self.merged_data = pd.merge(
                self.merged_data, df, on=on_column, how=how, suffixes=('_x', f'_{key}')
            )

    def calculate_final_grades(self, weights=None):
        """
        Calculates final grades for students using a weighted average, if weights are provided.

        Args:
            weights (dict, optional): A dictionary mapping grade component names to their weights.
                                      e.g., {'quiz': 0.2, 'exam': 0.5, 'project': 0.3}.
                                      If None, uses simple average.

        Returns:
            pd.DataFrame or None: DataFrame with final grades added, or None if data not merged.
        """
        if ...: #COMPLETE THIS: CHECK THAT DATASETS HAVE BEEN MERGED FIRST
            print("Data not merged. Please run merge_datasets first.")
            return None

        # Placeholder for implementation
        if weights:
             
            final_grades = 0
            for key, weight in weights.items():
                cols = [col for col in df.columns if key in col.lower()]
                if cols:
                    final_grades += ...  # COMPLETE THIS: Implement weighted average calculation
                else:
                    print(f"Warning: No columns found for '{key}'")
            df['final_grade'] = final_grades
        else:
            df['final_grade'] = ... # COMPLETE THIS: Implement simple average calculation

        self.merged_data = df
        return df  # return a new DataFrame with final grades

    def get_student_grades(self, student_id):
        """
        Retrieves all grade data for a specific student by their ID.

        Args:
            student_id (str or int): The ID of the student to retrieve grades for.

        Returns:
            pd.DataFrame or None: A DataFrame containing the student's grades, or None if not found or data not merged.
        """
        if ...: #COMPLETE THIS: CHECK THAT DATASETS HAVE BEEN MERGED FIRST
            print("Data not merged. Please run merge_datasets first.")
            return None
        
        # Normalize student_id in DataFrame to string zero-padded
        self.merged_data['student_id'] = self.merged_data['student_id'].astype(str).str.zfill(3)

        # Normalize the input student_id similarly
        student_id_str = str(student_id).zfill(3)
        
        return self.merged_data[self.merged_data['student_id'] == ...]   #COMPLETE THIS LINE!


&#9989; **Question 4.2 (6 points). Instantiate an object in this class and use it as follows:**
1. **Initialize a Gradebook Manager object using the three provided files `quiz_grades.csv`, `midterm.csv`, and `final_grades.csv`. (2 points)**
2. **Merge the three datasets. (1 point)**
3. **Compute the final grades in this class using the following weights: quizzes -> 35%, midterm -> 30%, and final exam -> 35%. (2 points)**
4. **Show the grades for the student with ID number `013`. (1 point)**
 

In [44]:
### Put your code here



---
### &#128721; STOP (1 point)
**Pause to commit your changes to your Git repository!**

Take a moment to save your notebook, commit  and push the changes to your Git repository.

---

## Part 5: Writing a simple Python code using different programming paradigms (17 points)

In this section, you will show your ability to use the three different programming paradigms we learned in class (linear, functional, and object-oriented). 

You work in a Chemistry laboratory. The lab prepares various chemical solutions with different solutes like NaCl, HCl, KOH, etc. Each solution is recorded with the following details: `Solution ID`, `Name`, `Mass(g)`, `Volume(mL)`, and `Molar mass(g/mol)`. 

Suppose you are handed the following chemical solutions data:
| solution_id | name | mass_g | volume_ml | molar_mass | 
| ----------- | ------ | ------ | --------- | ---------- |
| S001        | NaCl   | 5.8    | 100       | 58.44      |           
| S002        | HCl    | 3.65   | 250       | 36.46      |
| S003        | KOH    | 4.5    | 150       | 56.11      |

And you are asked to compute the following for each solution:

1. Record the data as a Pandas dataframe 
2. Moles of solute: mass_g / molar_mass
3. Volume in liters: volume_ml / 1000
4. Molarity (mol/L): moles / volume_L
5. Add the computed molarity as a new column in the DataFrame


&#9989; **Question 5.1 Write code for the above task using linear programming (5 points):** 

In [ ]:
### Put your code here

&#9989; **Question 5.2 Write code for the above task using functional programming (5 points):** 

In [ ]:
### Put your code here

&#9989; **Question 5.3 Write code for the above task using object-oriented programming (5 points):** 

In [ ]:
### Put your code here

&#9989; **Question 5.4 Based on these results, comment on the advantages and disadvantages between linear, functional and OOP programming (1 point):** 

<font size=+3>&#9998;</font> _Put your answer here_

---
### &#128721; STOP (1 point)
**Pause to commit your changes to your Git repository!**

Take a moment to save your notebook, commit  and push the changes to your Git repository.

---

## You're done! Congrats on finishing your CMSE 202 Midterm!

Make sure all of your changes to your repository are committed and pushed to GitHub. Also upload a copy of this notebook to the dropbox on D2L in case something went wrong with your repository or if you couldn't get the repository to work.